src/main/java/com/pusher/client/channel/impl/ChannelManager.java

```diff
@@ -137,6 +152,12 @@ public void run() {
 
                 if (connection.getState() == ConnectionState.CONNECTED) {
                     try {
+
+                        // only private encrypted channels can prepare themselves atm
+                        if (channel instanceof PrivateEncryptedChannelImpl) {
```

Comment:
I like this approach of separated prepare and teardown, but it also feels like private and presence have the same step, it's just not separated out there.

Rather than testing the type here, why don't we add `prepare` and `teardown` to the base `ChannelImpl` with no-op implementations and then override with the more specific behaviour for the non-public channel types?

Reply:
Totally agree, I didn't want to apply the same changes to the other channels without getting consensus this was a sensible thing to be adding. I'm not sure if that change should be made in the next week before e2e encryption is completed, but it can definitely come soon after as an improvement?